splice_machine/src/main/java/com/splicemachine/derby/vti/ORCRecordIterator.java

```diff
@@ -0,0 +1,246 @@
+import org.apache.hadoop.hive.ql.io.orc.Reader;
+import org.apache.hadoop.hive.ql.io.orc.RecordReader;
+import org.apache.hadoop.hive.serde2.objectinspector.StructObjectInspector;
+import org.apache.log4j.Logger;
+
+import com.splicemachine.db.iapi.sql.execute.ExecRow;
+import com.splicemachine.derby.impl.sql.execute.operations.LocatedRow;
+
+/**
+ * This is iterator for the records in the ORC files. It supports either a
+ * single file or Files in a folder. In case of file the next() will return the
+ * rows from one file, and then move to the next file
+ * 
+ * @author jramineni
+ * 
+ */
+
+public class ORCRecordIterator implements Iterable<LocatedRow>,
+Iterator<LocatedRow>, Closeable {
+
+	private static final Logger LOG = Logger.getLogger(ORCRecordIterator.class);
+
+	// The format of the Record to be returned
+	private ExecRow execRow;
+
+	// This is contains the details of the ORC File Record format
+	private StructObjectInspector inspector;
+
+	// Records in Orc File to be processed
+	RecordReader records;
+
+	// Current row in the ORC File to process
+	Object row = null;
+
+	// Indicates if a single file or files in folder are considered
+	boolean isDir = false;
+
+	// Iterator for files in a folder, when folder is specified
+	RemoteIterator<LocatedFileStatus> fileList = null;
+
+	// File System
+	FileSystem filesystem;
+
+	/**
+	 * Constructor performs initialization and sets the record iterator
+	 * 
+	 * @param filesystem
+	 *            : HDFS File System
+	 * @param filePath
+	 *            : File or folder path in HDFS
+	 * @param execRow
+	 *            : Format of the result record
+	 * @throws IOException
+	 */
+	public ORCRecordIterator(FileSystem filesystem, Path filePath,
+			ExecRow execRow) {
+		// set the instance variable of filesystem to use later
+		this.filesystem = filesystem;
+
+		// Set the instance variable of the result record format to be used
+		// later
+		this.execRow = execRow;
+
+		try {
+			Path curFiletoProcess = null;
+			Reader reader;
+
+			// Check if filePath specifes afile or folder
+			// If its folder, set the flag, and get the first file in the folder
+			if (filesystem.isDirectory(filePath)) {
+				isDir = true;
+				this.fileList = filesystem.listFiles(filePath, false);
+				curFiletoProcess = fileList.next().getPath();
+			} else {
+
+				curFiletoProcess = filePath;
+			}
+
+			// Get the reader for the single file (first file in case of folder)
+			reader = getReader(curFiletoProcess);
+
+			// Get the inspector for the format of the record in the ORC File
+			this.inspector = (StructObjectInspector) reader
+					.getObjectInspector();
+
+			// Retrieve the Records from reader to process
+			records = reader.rows();
+
+		} catch (Exception e) {
+			try {
+				if (records != null)
+					records.close();
+			} catch (Exception cE) {
+				throw new RuntimeException(cE);
+			}
+			throw new RuntimeException(e);
+		}
+
+	}
+
+	@Override
+	public Iterator<LocatedRow> iterator() {
+		return this;
```

Comment:
This is an artifact of Spark's API that imposes returning Iterables where only Iterators where needed. We do this (implementing Iterable and Iterator on the same object) all over the place. It will go away when we move to Spark 2.0 (hopefully today).